contrib/etime/etime.c

```diff
@@ -0,0 +1,239 @@
+
+static ExecutorStart_hook_type prev_ExecutorStart_hook = NULL;
+static ExecutorEnd_hook_type prev_ExecutorEnd_hook = NULL;
+
+/*
+ * An ETimeEvent records the content and execution time of a query
+ */
+typedef struct
+{
+	char *queryText;						/* sql text */
+	TimestampTz startTime;			/* start time the query executing */
+	MemoryContext queryContext; /* Context for query tracking rows */
+} ETimeEvent;
+
+/*
+ * A simple FIFO queue to keep track of the current stack of etime events.
+ */
+typedef struct ETimeEventStackItem
+{
+	struct ETimeEventStackItem *next;
+
+	ETimeEvent eTimeEvent;
+
+	MemoryContext context;
+	MemoryContextCallback contextCallback;
+} ETimeEventStackItem;
+
+ETimeEventStackItem *eTimeEventStack = NULL;
+
+/*
+ * Respond to callbacks registered with MemoryContextRegisterResetCallback().
+ * Removes the event(s) off the stack that have become obsolete once the
+ * MemoryContext has been freed.  The callback should always be freeing the top
+ * of the stack, but the code is tolerant of out-of-order callbacks.
+ */
+static void
+stack_free(void *stackFree)
+{
+	ETimeEventStackItem *nextItem = eTimeEventStack;
+
+	/* Only process if the stack contains items */
+	while (nextItem != NULL)
+	{
+		/* Check if this item matches the item to be freed */
+		if (nextItem == (ETimeEventStackItem *)stackFree)
+		{
+			/* Move top of stack to the item after the freed item */
+			eTimeEventStack = nextItem->next;
+			return;
+		}
+
+		nextItem = nextItem->next;
+	}
+}
+
+/*
+ * Push a new etime event onto the stack and create a new memory context to
+ * store it.
+ */
+static ETimeEventStackItem *
+stack_push()
+{
+	MemoryContext context;
+	MemoryContext contextOld;
+	ETimeEventStackItem *stackItem;
+
+	/*
+	 * Create a new memory context to contain the stack item.  This will be
+	 * free'd on stack_pop, or by our callback when the parent context is
+	 * destroyed.
+	 */
+	context = AllocSetContextCreate(CurrentMemoryContext,
+																	"etime stack context",
+																	ALLOCSET_DEFAULT_SIZES);
+
+	/* Save the old context to switch back to at the end */
+	contextOld = MemoryContextSwitchTo(context);
+
+	/* Create our new stack item in our context */
+	stackItem = palloc0(sizeof(ETimeEventStackItem));
+	stackItem->context = context;
+
+	/*
+	 * Setup a callback in case an error happens.  stack_free() will truncate
+	 * the stack at this item.
+	 */
+	stackItem->contextCallback.func = stack_free;
+	stackItem->contextCallback.arg = (void *)stackItem;
+	MemoryContextRegisterResetCallback(context,
+																		 &stackItem->contextCallback);
+
+	/* Push new item onto the stack */
+	if (eTimeEventStack != NULL)
+		stackItem->next = eTimeEventStack;
+	else
+		stackItem->next = NULL;
+
+	eTimeEventStack = stackItem;
+
+	MemoryContextSwitchTo(contextOld);
+
+	return stackItem;
+}
+
+/**
+ * Hook ExecutorStart to get the query text and its execution time
+ */
+static void
+my_ExecutorStart_hook(QueryDesc *queryDesc, int eflags)
```

Comment:
It seems that this plugin does not support accessing the queries whose execution time above the threshold as a table or a view by the way of SQL but not log files. What about implement it referring to the existing plugin pg_stat_log?

Reply:
I am sorry I forget that pg_stat_log is only available in the commercial version currently. In short, it support a set of functions to record, show and reset the statement statistics, including the slow queries and its begin time, end time, total time and so on. It is more convenient for users to check them by accessing a view-like function but not the log files directly.